### PR TITLE
Actually fix major sync detection

### PR DESCRIPTION
**Pull request by nazar-pc**
https://github.com/paritytech/substrate/pull/11547 didn't actually resolve https://github.com/paritytech/substrate/issues/11510.

`median > self.best_queued_number` caused `best_seen` to return `None` way before there is just 5 blocks left to import, causing `SyncStatus::status` flip to `SyncState::Idle` prematurely and changing `is_major_sync` to `false` as the result even though there might still be over 2k blocks left to import (queue for blocks to import has size 2048).

The solution is to check `median` for `SyncStatus::status` calculation, but leave `best_seen_block` as is (as I'm not 100% of implications of changing it).

Before:
<details>

```
2022-08-26 01:28:40 [PrimaryChain] Is major syncing: true    
2022-08-26 01:28:41 [PrimaryChain] Is major syncing: true    
2022-08-26 01:28:41 [PrimaryChain] ⚙️  Syncing 186.2 bps, target=#16494 (2 peers), best: #14266 (0x1f1f…43db), finalized #14166 (0x9dc2…9893), ⬇ 2.1MiB/s ⬆ 1.4kiB/s    
2022-08-26 01:28:42 [PrimaryChain] Is major syncing: true    
2022-08-26 01:28:43 [PrimaryChain] Is major syncing: false    
2022-08-26 01:28:44 [PrimaryChain] Is major syncing: false    
2022-08-26 01:28:45 [PrimaryChain] Is major syncing: false    
2022-08-26 01:28:46 [PrimaryChain] Is major syncing: false    
2022-08-26 01:28:46 [PrimaryChain] 💤 Idle (2 peers), best: #15065 (0x882a…bac5), finalized #14964 (0x4f5e…4989), ⬇ 1.3MiB/s ⬆ 2.2kiB/s    
2022-08-26 01:28:47 [PrimaryChain] Is major syncing: false    
2022-08-26 01:28:48 [PrimaryChain] Is major syncing: false    
2022-08-26 01:28:49 [PrimaryChain] Is major syncing: false    
2022-08-26 01:28:49 [PrimaryChain] ✨ Imported #15544 (0x1726…3a69)    
2022-08-26 01:28:49 [PrimaryChain] ✨ Imported #15545 (0xadd3…f7c2)    
```

</details>

After:
<details>

```
2022-08-26 02:23:00 [PrimaryChain] Is major syncing: true    
2022-08-26 02:23:01 [PrimaryChain] Is major syncing: true    
2022-08-26 02:23:01 [PrimaryChain] ⚙️  Preparing 62.0 bps (2 peers), best: #17668 (0xbf6a…f5a4), finalized #17567 (0xd79e…fe55), ⬇ 55.9kiB/s ⬆ 0.2kiB/s    
2022-08-26 02:23:02 [PrimaryChain] Is major syncing: true    
2022-08-26 02:23:03 [PrimaryChain] Is major syncing: true    
2022-08-26 02:23:04 [PrimaryChain] Is major syncing: true    
2022-08-26 02:23:05 [PrimaryChain] Is major syncing: true    
2022-08-26 02:23:06 [PrimaryChain] Is major syncing: true    
2022-08-26 02:23:06 [PrimaryChain] ⚙️  Syncing 63.2 bps, target=#18363 (2 peers), best: #17984 (0x8377…7c7e), finalized #17884 (0x8d2b…693c), ⬇ 2.2MiB/s ⬆ 1.4kiB/s    
2022-08-26 02:23:07 [PrimaryChain] Is major syncing: true    
2022-08-26 02:23:08 [PrimaryChain] Is major syncing: true    
2022-08-26 02:23:09 [PrimaryChain] Is major syncing: true    
2022-08-26 02:23:10 [PrimaryChain] Is major syncing: true    
2022-08-26 02:23:11 [PrimaryChain] Is major syncing: true    
2022-08-26 02:23:11 [PrimaryChain] 💤 Idle (0 peers), best: #18246 (0x7d61…fd5b), finalized #18145 (0x25c3…97b2), ⬇ 211.6kiB/s ⬆ 0.2kiB/s    
2022-08-26 02:23:12 [PrimaryChain] Is major syncing: true    
2022-08-26 02:23:13 [PrimaryChain] Is major syncing: true    
2022-08-26 02:23:14 [PrimaryChain] Is major syncing: true    
2022-08-26 02:23:15 [PrimaryChain] Is major syncing: true    
2022-08-26 02:23:16 [PrimaryChain] Is major syncing: true    
2022-08-26 02:23:16 [PrimaryChain] 💤 Idle (0 peers), best: #18362 (0x5f93…0ff3), finalized #18262 (0xb835…2688), ⬇ 0.2kiB/s ⬆ 0.2kiB/s    
2022-08-26 02:23:16 [PrimaryChain] ✨ Imported #18363 (0xc01f…8821)    
2022-08-26 02:23:16 [PrimaryChain] ✨ Imported #18364 (0xfe71…bbda)    
2022-08-26 02:23:16 [PrimaryChain] ✨ Imported #18365 (0xf875…4ca1)    
2022-08-26 02:23:17 [PrimaryChain] Is major syncing: false    
2022-08-26 02:23:17 [PrimaryChain] ✨ Imported #18366 (0xe97d…d925)    
2022-08-26 02:23:18 [PrimaryChain] Is major syncing: false    
```

</details>

Polkadot address: 1vSxzbyz2cJREAuVWjhXUT1ds8vBzoxn2w4asNpusQKwjJd

### CODE REVIEW

**Comment by nazar-pc** [comment]
cc @bkchr @andresilva and @arkpar as authors/reviewers of mentioned PR.

**Comment by nazar-pc** [comment]
Hm... turns out it was actually un-fixed by https://github.com/paritytech/substrate/pull/11817

**Comment by bkchr** [comment]
Maybe we should just introduce a third state `Importing` when we are finished with downloading, but still need to import the blocks?

**Comment by nazar-pc** [comment]
Sounds reasonable to me, I can do that too

**Comment by bkchr** [comment]
> Sounds reasonable to me, I can do that too

Would be nice. Block production would not start until we are in IDLE state. However, internally we can distinguish between the different states.

**Comment by nazar-pc** [comment]
Added `SyncState::Importing`, but it now looks awkward with `best_seen_block` being separate. I feel like `SyncState` enum should contain that number inside of it, then it would be easier to handle things in informant. WDYT?

File: client/network/common/src/sync.rs [hunk 1]
```diff
@@ -67,6 +69,8 @@ pub struct SyncStatus<Block: BlockT> {
 	pub state: SyncState,
 	/// Target sync block number.
 	pub best_seen_block: Option<NumberFor<Block>>,
+	/// Are we actively catching up with the chain?
+	pub is_major_syncing: bool,
```
Comment (arkpar) [inline client/network/common/src/sync.rs:73]: This could be a method of `SyncState` instead.

**Comment by arkpar** [comment]
> Added SyncState::Importing, but it now looks awkward with best_seen_block being separate. I feel like SyncState enum should contain that number inside of it, then it would be easier to handle things in informant. WDYT?

Agree. Both `Importing` and `Downloading` should have it though.

**Comment by nazar-pc** [comment]
Pushed corresponding changes, but now I'm wondering if `best_seen_block` should be `Some(target)` in case of `SyncState::Importing`. It wasn't before, but maybe it should?

**Comment by arkpar** [comment]
Do you mean `NetworkStatus::best_seen_block`? I suggest it should be removed alltogeter, since we already have the target block in `NetworkStatus::sync_state`

File: client/network/src/protocol.rs [hunk 2]
```diff
@@ -501,7 +501,11 @@ where
 
 	/// Target sync block number.
 	pub fn best_seen_block(&self) -> Option<NumberFor<B>> {
-		self.chain_sync.status().best_seen_block
+		match self.chain_sync.status().state {
+			SyncState::Idle => None,
+			SyncState::Downloading { target } => Some(target),
+			SyncState::Importing { .. } => None,
```
Comment (arkpar) [inline client/network/src/protocol.rs:507]: Could be `Some(target)` here indeed.

File: client/network/common/src/sync.rs [hunk 3]
```diff
@@ -44,11 +44,27 @@ pub struct PeerInfo<Block: BlockT> {
...
+	/// All blocks are downloaded and are being imported.
+	Importing { target: BlockNumber },
+}
+
+impl<BlockNumber> SyncState<BlockNumber>
+where
+	BlockNumber: Clone,
+{
+	/// Are we actively catching up with the chain? If so, returns `Some(target)`.
+	pub fn major_sync_target(&self) -> Option<BlockNumber> {
```
Comment (bkchr) [inline client/network/common/src/sync.rs:61]: Could we please add one `is_major_syncing`?

File: client/network/sync/src/lib.rs [hunk 4]
```diff
@@ -1729,12 +1734,7 @@ where
 			let middle = best_seens.len() / 2;
 
 			// Not the "perfect median" when we have an even number of peers.
-			let median = *best_seens.select_nth_unstable(middle).1;
-			if median > self.best_queued_number {
-				Some(median)
-			} else {
-				None
-			}
```
Comment (bkchr) [inline client/network/sync/src/lib.rs:1737]: This reverts: https://github.com/paritytech/substrate/pull/12114

Comment (bkchr) [reply client/network/sync/src/lib.rs:1737]: As far as I can see, we would run into the same problem with your code.

**Review by bkchr — COMMENTED** [review]
Besides your changes, adding some test would be really nice.

File: client/network/sync/src/lib.rs (hunk 4 above)
Comment (nazar-pc) [reply client/network/sync/src/lib.rs:1737]: I think you meant #11817, right? If so then answer to my previous question about `SyncState::Importing` was no, it shouldn't return `Some(target)`, I'll revert it back to the previous logic as of #11817.

**Review by bkchr — APPROVED** [review]

**Comment by bkchr** [comment]
bot merge

**Comment by bkchr** [comment]
bot merge

**Comment by nazar-pc** [comment]
I see a sync-related test failing, I'll look into it

**Comment by bkchr** [comment]
> I see a sync-related test failing, I'll look into it

Any update on this? @nazar-pc

**Comment by nazar-pc** [comment]
I'll get back to this soon, sorry it has been a while

**Comment by nazar-pc** [comment]
Last commit with optimization wasn't correct, so I just reverted it.
Merge conflict was due to import, which is easy.
Last commit is because otherwise `sc-network` doesn't compile (probably works in CI due to feature unification with some other crate).

**Comment by bkchr** [comment]
bot merge

**Comment by bkchr** [comment]
bot rebase

**Comment by bkchr** [comment]
bot merge

**Pull request merged**